test/e2e/volume_provisioning.go

```diff
@@ -108,61 +216,182 @@ var _ = framework.KubeDescribe("Dynamic Provisioning", func() {
+				[]string{"aws"},
+				"kubernetes.io/aws-ebs",
+				map[string]string{
+					"encrypted": "true",
+				},
+				"1Gi",
+				"1Gi",
+				func(volume *v1.PersistentVolume) error {
+					return checkAWSEBS(volume, "gp2", true)
+				},
+			},
+			// OpenStack generic tests (works on all OpenStack deployments)
+			{
+				"should provision generic Cinder volume on OpenStack",
+				[]string{"openstack"},
+				"kubernetes.io/cinder",
+				map[string]string{},
+				"1.5Gi",
+				"2Gi",
+				nil, // there is currently nothing to check on OpenStack
+			},
+			{
+				"should provision Cinder volume with empty volume type and zone on OpenStack",
+				[]string{"openstack"},
+				"kubernetes.io/cinder",
+				map[string]string{
+					"type":         "",
+					"availability": "",
+				},
+				"1.5Gi",
+				"2Gi",
+				nil, // there is currently nothing to check on OpenStack
+			},
+			// vSphere generic test
+			{
+				"should provision generic vSphere",
+				[]string{"vsphere"},
+				"kubernetes.io/vsphere-volume",
+				map[string]string{},
+				"1.5Gi",
+				"1.5Gi",
+				nil,
+			},
+		}
 
-	framework.KubeDescribe("DynamicProvisioner Beta", func() {
-		It("should create and delete persistent volumes [Slow] [Volume]", func() {
-			framework.SkipUnlessProviderIs("openstack", "gce", "aws", "gke")
+		var betaTest *storageClassTest
+		for i, t := range tests {
+			// Beware of clojure, use local variables instead of those from
+			// outer scope
+			test := t
+			suffix := fmt.Sprintf("%d", i)
+			It(test.name+" [Slow] [Volume]", func() {
+				if len(t.cloudProviders) > 0 {
+					framework.SkipUnlessProviderIs(test.cloudProviders...)
+				}
 
-			By("creating a StorageClass")
-			class := newBetaStorageClass("", "beta")
-			class, err := c.StorageV1beta1().StorageClasses().Create(class)
-			defer deleteStorageClass(c, class.Name)
-			Expect(err).NotTo(HaveOccurred())
+				// Remember the last supported test for subsequent test of beta API
+				betaTest = &test
 
-			By("creating a claim with a dynamic provisioning annotation")
-			claim := newClaim(ns)
-			claim.Annotations = map[string]string{
-				v1.BetaStorageClassAnnotation: class.Name,
-			}
+				class := newStorageClass(test, ns, suffix)
+				claim := newClaim(test, ns, suffix)
+				claim.Spec.StorageClassName = &class.Name
+				testDynamicProvisioning(test, c, claim, class)
+			})
+		}
 
-			defer func() {
-				framework.DeletePersistentVolumeClaim(c, claim.Name, ns)
-			}()
-			claim, err = c.Core().PersistentVolumeClaims(ns).Create(claim)
-			Expect(err).NotTo(HaveOccurred())
+		// Run the last test with storage.k8s.io/v1beta1 and beta annotation on pvc
+		if betaTest != nil {
+			It("Beta "+betaTest.name+" [Slow] [Volume]", func() {
+				class := newBetaStorageClass(*betaTest, "beta")
+				// we need to create the class manually, testDynamicProvisioning does not accept beta class
+				class, err := c.StorageV1beta1().StorageClasses().Create(class)
+				defer deleteStorageClass(c, class.Name)
```

Comment:
This works since `deleteStorageClass()` handles `apierrs.IsNotExists`.  Jan is there a reason why you chose to do a deferred delete before the resource is created? I'd like to better understand the thinking here.